### PR TITLE
HAI-1526 Send application invitations

### PR DESCRIPTION
**Pull request by pitkni**
# Description

Send hanke invitations to new hanke contacts and send application notification to every contact. Does not send messages to the inviter himself.

Additional: Fix email text versions so that variables are not converted to html escaped characters.

## Type of change

- [ ] Bug fix 
- [x] New feature 
- [ ] Other

# Instructions for testing
Create johtoselvitys with contacts. Send application. Check localhost:3003 that emails are sent correctly, i.e. new contacts receive hanke invitation, all contacts receive application notification.

# Checklist:

- [x] I have written new tests (if applicable)
- [x] I have ran the tests myself (if applicable)
- [ ] I have made necessary changes to the documentation, link to confluence
 or other location: 

# Other relevant info
Please describe here if there is e.g. some requirements for this change or
 other info that the tester/user needs to know.

### CODE REVIEW

File: services/hanke-service/src/main/kotlin/fi/hel/haitaton/hanke/permissions/HankeKayttajaService.kt [hunk 1]
```diff
@@ -1,21 +1,42 @@
...
+import fi.hel.haitaton.hanke.typedContacts
+import fi.hel.haitaton.hanke.userContact
 import java.util.UUID
 import mu.KotlinLogging
 import org.springframework.stereotype.Service
 import org.springframework.transaction.annotation.Transactional
 
 private val logger = KotlinLogging.logger {}
 
+interface UserContact {
```
Comment (corvidian) [inline services/hanke-service/src/main/kotlin/fi/hel/haitaton/hanke/permissions/HankeKayttajaService.kt:27]: This could be declared in a separate file. That file could also contain the two implementations and the stuff that's been added to Utils.kt.

This could be a sealed interface.

File: services/hanke-service/src/main/kotlin/fi/hel/haitaton/hanke/Utils.kt [hunk 2]
```diff
@@ -76,3 +86,43 @@ fun BusinessId.isValidBusinessId(): Boolean {
 }
 
 fun List<CustomerWithContacts>.ordererCount() = flatMap { it.contacts }.count { it.orderer }
+
+fun userContact(name: String?, email: String?): HankeUserContact? {
```
Comment (corvidian) [inline services/hanke-service/src/main/kotlin/fi/hel/haitaton/hanke/Utils.kt:90]: This could use an expression body like the other one.

File: services/hanke-service/src/main/kotlin/fi/hel/haitaton/hanke/permissions/HankeKayttajaService.kt [hunk 3]
```diff
@@ -280,28 +312,64 @@ class HankeKayttajaService(
...
+        hankeTunnus: String,
+        hankeNimi: String,
+        inviter: HankeKayttajaEntity?,
         recipient: HankeKayttajaEntity,
     ) {
-        logger.info { "Sending Hanke invitations." }
+        logger.info { "Sending Hanke invitation." }
+
+        if (inviter == null) {
+            logger.warn { "Inviter kaytaja null, will not send Hanke invitation." }
```
Comment (corvidian) [inline services/hanke-service/src/main/kotlin/fi/hel/haitaton/hanke/permissions/HankeKayttajaService.kt:324]:
```suggestion
            logger.warn { "Inviter kayttaja null, will not send Hanke invitation." }
```

File: services/hanke-service/src/main/kotlin/fi/hel/haitaton/hanke/permissions/HankeKayttajaService.kt [hunk 4]
```diff
@@ -50,21 +59,37 @@ class HankeKayttajaService(
...
+
+        val inviter = getKayttajaByUserId(hankeId, userId)
+        val contacts = application.applicationData.typedContacts().removeInviter(inviter)
 
         filterNewContacts(hankeId, contacts).forEach { contact ->
-            createTunnisteAndKayttaja(hankeId, contact, userId)
+            createTunnisteAndKayttaja(hankeId, hankeTunnus, hankeNimi, inviter, contact, userId)
+        }
+        contacts.forEach { contact ->
+            sendApplicationNotification(
```
Comment (corvidian) [inline services/hanke-service/src/main/kotlin/fi/hel/haitaton/hanke/permissions/HankeKayttajaService.kt:86]: Unlike the invitations, these emails are not related to saving new HankeKayttaja. I don't think this should even be in this service.

File: services/hanke-service/src/main/kotlin/fi/hel/haitaton/hanke/application/ApplicationService.kt [hunk 5]
```diff
@@ -218,16 +217,41 @@ open class ApplicationService(
...
+    ) {
+        val hankeId = hanke.id
+        val hankeTunnus = hanke.hankeTunnus
+        val hankeNimi = hanke.nimi
+
+        if (hankeId == null || hankeTunnus == null || hankeNimi == null) {
+            throw ApplicationArgumentException(
+                "Hanke info missing: (id=$hankeId, tunnus=$hankeTunnus, nimi=$hankeNimi)"
+            )
+        }
```
Comment (corvidian) [inline services/hanke-service/src/main/kotlin/fi/hel/haitaton/hanke/application/ApplicationService.kt:244]: We should push to get HAI-1893 to the next sprint to avoid this kind of nonsense.

File: services/hanke-service/src/main/kotlin/fi/hel/haitaton/hanke/Utils.kt (hunk 2 above)
Comment (pitkni) [reply services/hanke-service/src/main/kotlin/fi/hel/haitaton/hanke/Utils.kt]: Fixed.

File: services/hanke-service/src/main/kotlin/fi/hel/haitaton/hanke/permissions/HankeKayttajaService.kt (hunk 1 above)
Comment (pitkni) [reply services/hanke-service/src/main/kotlin/fi/hel/haitaton/hanke/permissions/HankeKayttajaService.kt]: Good suggestions, fixed.

File: services/hanke-service/src/main/kotlin/fi/hel/haitaton/hanke/permissions/HankeKayttajaService.kt (hunk 4 above)
Comment (pitkni) [reply services/hanke-service/src/main/kotlin/fi/hel/haitaton/hanke/permissions/HankeKayttajaService.kt:86]: It was placed here as it has a lot of similarities to hanke invitations. But you are making a good point here. HankeKayttajaService should not know anything about application notifications. I'll move that part to application service.

File: services/hanke-service/src/main/kotlin/fi/hel/haitaton/hanke/permissions/HankeKayttajaService.kt [hunk 6]
```diff
@@ -280,28 +312,64 @@ class HankeKayttajaService(
...
-                hankeNimi = hanke.nimi!!,
+                hankeTunnus = hankeTunnus,
+                hankeNimi = hankeNimi,
                 invitationToken = recipient.kayttajaTunniste!!.tunniste,
             )
         )
     }
 
-    private fun createTunniste(hankeId: Int, userId: String): KayttajaTunnisteEntity {
+    private fun sendApplicationInvitation(
```
[the remark this answers is not in the document]
Comment (pitkni) [reply services/hanke-service/src/main/kotlin/fi/hel/haitaton/hanke/permissions/HankeKayttajaService.kt]: Renamed.

File: services/hanke-service/src/main/kotlin/fi/hel/haitaton/hanke/permissions/HankeKayttajaService.kt (hunk 4 above)
Comment (pitkni) [reply services/hanke-service/src/main/kotlin/fi/hel/haitaton/hanke/permissions/HankeKayttajaService.kt]: Done.

File: services/hanke-service/src/main/kotlin/fi/hel/haitaton/hanke/permissions/HankeKayttajaService.kt (hunk 1 above)
Comment (corvidian) [reply services/hanke-service/src/main/kotlin/fi/hel/haitaton/hanke/permissions/HankeKayttajaService.kt]: The extensions can be moved to the same file as well. I think Utils.kt should be reserved for highly generic things. Like things that are not directly related to any domain objects.

File: services/hanke-service/src/main/kotlin/fi/hel/haitaton/hanke/permissions/HankeKayttajaService.kt (hunk 6 above)
[the remark this answers is not in the document]
Comment (corvidian) [reply services/hanke-service/src/main/kotlin/fi/hel/haitaton/hanke/permissions/HankeKayttajaService.kt]: Should things like `EmailTemplate.INVITATION_APPLICATION` and `inviterName` be renamed as well?

File: services/hanke-service/src/main/kotlin/fi/hel/haitaton/hanke/permissions/HankeKayttajaService.kt (hunk 1 above)
Comment (pitkni) [reply services/hanke-service/src/main/kotlin/fi/hel/haitaton/hanke/permissions/HankeKayttajaService.kt]: Done. Also mover orderercount to applicationdata file.

File: services/hanke-service/src/main/kotlin/fi/hel/haitaton/hanke/permissions/HankeKayttajaService.kt (hunk 6 above)
[the remark this answers is not in the document]
Comment (pitkni) [reply services/hanke-service/src/main/kotlin/fi/hel/haitaton/hanke/permissions/HankeKayttajaService.kt]: Yes. Do you have suggestions for new name for inviterName? applicationSender?

[the remark this answers is not in the document]
Comment (pitkni) [reply services/hanke-service/src/main/kotlin/fi/hel/haitaton/hanke/permissions/HankeKayttajaService.kt]: I'll use senderName as it shorter and it can be thought of as the mail sender and application sender (to allu).

[the remark this answers is not in the document]
Comment (pitkni) [reply services/hanke-service/src/main/kotlin/fi/hel/haitaton/hanke/permissions/HankeKayttajaService.kt]: Done.

**Review by corvidian — APPROVED** [review]
